renderdoc/driver/gl/gl_pixelhistory.cpp

```diff
@@ -276,13 +383,226 @@ void QueryFailedTests(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+                         return h1.eventId < h2.eventId;
+                       });
+  RDCASSERT(historyIndex != history.end());
+
+  if(queryType == PerFragmentQueryType::PostMod)
+  {
+    if(historyIndex != history.begin())
+    {
+      --historyIndex;
+
+      // Because we are replaying this draw into our own framebuffer
+      // We need to set the values to the ones that they would be
+      // before this draw call.
+      driver->glClearColor(
+          historyIndex->postMod.col.floatValue[0], historyIndex->postMod.col.floatValue[1],
+          historyIndex->postMod.col.floatValue[2], historyIndex->postMod.col.floatValue[3]);
+
+      GLboolean depthWriteMask = false;
+      driver->glGetBooleanv(eGL_DEPTH_WRITEMASK, &depthWriteMask);
+      driver->glDepthMask(true);
+      driver->glClearDepth(historyIndex->postMod.depth);
+
+      driver->glClear(eGL_COLOR_BUFFER_BIT | eGL_DEPTH_BUFFER_BIT);
+
+      driver->glDepthMask(depthWriteMask);
+      ++historyIndex;
+    }
+  }
+  else if(queryType == PerFragmentQueryType::PrimitiveId)
+  {
+    // we expect this value to be overwritten by the primitive id shader. if not
+    // it will cause an assertion that all color values are the same to fail
+    driver->glClearColor(0.84f, 0.17f, 0.2f, 0.49f);
+    driver->glClear(eGL_COLOR_BUFFER_BIT);
+
+    GLint currentProgram;
+
+    driver->glGetIntegerv(eGL_CURRENT_PROGRAM, &currentProgram);
+
+    driver->glUseProgram(GetPrimitiveIdProgram(driver, replay, resources, currentProgram));
+  }
+
+  for(size_t j = 0; j < std::max(numFragments, 1u); ++j)
+  {
+    ModificationValue modValue;
+    PixelValue pixelValue;
+    modValue.stencil = 0;
+    driver->glStencilFunc(eGL_EQUAL, (int)j, 0xff);
+    driver->glClear(eGL_STENCIL_BUFFER_BIT);
+
+    driver->ReplayLog(modEvent.eventId, modEvent.eventId, eReplay_Full);
+
+    if(queryType == PerFragmentQueryType::PrimitiveId)
+    {
+      int primitiveIds[4];
+      driver->glReadPixels(x, y, 1, 1, eGL_RGBA, eGL_FLOAT, (void *)primitiveIds);
+
+      RDCASSERT(primitiveIds[0] == primitiveIds[1] && primitiveIds[0] == primitiveIds[2] &&
+                primitiveIds[0] == primitiveIds[3]);
+      historyIndex->primitiveID = primitiveIds[0];
+    }
+    else
+    {
+      driver->glReadPixels(x, y, 1, 1, eGL_RGBA, eGL_FLOAT, (void *)pixelValue.floatValue.data());
+      driver->glReadPixels(x, y, 1, 1, eGL_DEPTH_COMPONENT, eGL_FLOAT, (void *)&modValue.depth);
+
+      modValue.col = pixelValue;
+
+      if(queryType == PerFragmentQueryType::ShaderOut)
+      {
+        historyIndex->shaderOut = modValue;
+      }
+      else if(queryType == PerFragmentQueryType::PostMod)
+      {
+        historyIndex->postMod = modValue;
+      }
+    }
+    ++historyIndex;
+  }
+}
+
+void WriteClearValues(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+                      const EventUsage clearEvent, int x, int y, rdcarray<PixelModification> &history)
```

Comment:
Why do you need a separate path for fetching postmod values for clears? Ideally it should be possible to fetch all postmods (except for postmods between different fragments in the same event) in a single pass while calculating per-event overdraw counts. This also means that postmods are fetched at the correct precision, since we want to correctly represent limited precision depending on the target format.

Reply:
Do we care about reading back stencil values for postmod? Otherwise I agree that we should move this all in a single post mod pass.

Reply:
For after each _event_ then yes, we want stencil values and should be able to obtain them along with depth and colour when doing a single pass (not changing the draws at all). Stencil values between different fragments of a draw we can't get with the current system because we're using stencil for fragment counting, that's a limitation we have to accept.